### PR TITLE
Print env vars when running with --debug

### PR DESCRIPTION
**Pull request by autoantwort**
Hopefully helps to debug https://discord.com/channels/400588936151433218/687365466422902841/959446195061792829

### CODE REVIEW

File: src/vcpkg/base/system.cpp
Comment (BillyONeal) [inline src/vcpkg/base/system.cpp:18]:
```suggestion
#if !defined(_WIN32)
extern char** environ;
#endif
```

File: src/vcpkg/base/system.cpp
Comment (BillyONeal) [inline src/vcpkg/base/system.cpp:235]: I have a mild preference for a bespoke struct with a dtor over unique_ptr since that introduces unnecessary null checks etc.

File: src/vcpkg/base/system.cpp
Comment (BillyONeal) [inline src/vcpkg/base/system.cpp:245]:
```suggestion
        for (LPWCH i = env_block.get();;)
        {
            auto this_length = strlen(i);
            if (this_length == 0) { break; }
            msg::write_unlocalized_text_to_stdout(Color::none, Strings::concat(Strings::to_utf8(i, this_length), '\n'));
            i += this_length + 1;
        }
```

I would like to avoid so many repeated strlens :)

File: src/vcpkg/base/system.cpp
Comment (BillyONeal) [inline src/vcpkg/base/system.cpp:230]:
```suggestion
    std::string get_environment_block()
```

and let the calling debug place do the write_unlocalized_Xxx?

That way we make it clear that the reason this is unlocalized is because it's behind the debugging flag.

**Review by BillyONeal — CHANGES_REQUESTED** [review]
Do we actually want to do this given that the resulting reports are likely to contain PII?

**Comment by ras0219-msft** [comment]
I think having a separate `--debug-env` makes sense for this -- the environment block is definitely important for debugging, but it also is very large and potentially sensitive. Specifically:
- `--debug` without `--debug-env` should print a single line saying "To include the environment block in debug output, pass `--debug-env`"
- `--debug` with `--debug-env` causes that single line to be replaced with the full environment dump
- `--debug-env` implies `--debug`, if it is passed alone. At some future point, one could imagine splitting off other debug information into various other switches.

File: src/vcpkg/base/system.cpp
Comment (autoantwort) [reply src/vcpkg/base/system.cpp:245]: Hm if an variable has an empty string as value, this would break the loop.

Comment (BillyONeal) [reply src/vcpkg/base/system.cpp:245]: @autoantwort The behavior isn't changed from the original:

```
            while (*i != L'\0')
            {
                ++i;
            }
```

is `i += strlen(i)`.

and `*i != L'\0'` will stop on an empty string.

As far as I know the entries are going to be of the form `KEY=VALUE\0` though so I don't think it actually occurs?

Comment (autoantwort) [reply src/vcpkg/base/system.cpp:245]: True my mistake. It was too long ago and I made some wrong assumptions while writing my answer 😅

File: src/vcpkg/base/system.cpp
Comment (autoantwort) [reply src/vcpkg/base/system.cpp:235]: I don't think that this one null check is really important ...

File: src/vcpkg/base/system.cpp
Comment (autoantwort) [reply src/vcpkg/base/system.cpp:18]: We need `#include <processenv.h>` for mingw

Comment (BillyONeal) [reply src/vcpkg/base/system.cpp:18]: Hmmm can you add a comment to that effect then?

File: src/vcpkg/base/system.cpp
Comment (BillyONeal) [reply src/vcpkg/base/system.cpp:236]: IIRC it's more than one but it isn't enough to lose sleep over, which is why I pointed out the problem rather than saying please change it.

However:

```
struct EnvironmentStringsW {
    LPWCH strings;
    EnvironmentStringsW() : strings(GetEnvironmentStringsW()) {}
    ~EnvironmentStringsW() { FreeEnvironmentStringsW(strings); }
    EnvironmentStringsW(const EnvironmentStringsW&) = delete;
};
```

or similar isn't really much more code than you wrote here (and avoids shadowing the name free, etc.)

I'm not blocking over this suggestion though.

**Review by vicroms — APPROVED** [review]

File: src/vcpkg.cpp
Comment (ras0219-msft) [inline src/vcpkg.cpp:266]:
```suggestion
                                              "[DEBUG] The following environment variables are currently set:\n" +
```

**Review by ras0219-msft — APPROVED** [review]

**Review by ras0219-msft — APPROVED** [review]
I noticed this was using `strlen(i)` on the `wchar_t*` environment block; I'm not sure how that compiled, but I wanted to replace that with `wcslen()` anyway. Since I was in the area, I applied @BillyONeal's suggestion above as well.

This LGTM, but I'd appreciate a second approval from another vcpkg maintainer before merging.

File: src/vcpkg/base/system.cpp
Comment (BillyONeal) [inline src/vcpkg/base/system.cpp:253]:
```suggestion
            result.append(*s).push_back('\n');
```

For consistency with the Windows version.

File: src/vcpkg/base/system.cpp
Comment (BillyONeal) [inline src/vcpkg/base/system.cpp:247]:
```suggestion
        for (LPWCH i = env_block.strings;;)
        {
            const auto len = wcslen(i);
            if (len == 0) { break; }
            result.append(Strings::to_utf8(i, len)).push_back('\n');
            i += len + 1;
        }
```

IMO I think infinite loop is slightly less 😬 than assignment-in-conditional-expression like that (and also avoids leaking len) but this is an extremely minor nitpick. Feel free to disregard.

**Review by BillyONeal — APPROVED** [review]

**Pull request merged**
